src/libfetchers/unix/git.cc

```diff
@@ -167,11 +167,16 @@ struct GitInputScheme : InputScheme
 {
     std::optional<Input> inputFromURL(const ParsedURL & url, bool requireTree) const override
     {
+        if (url.scheme != "git" && url.scheme.compare(0, 4, "git+") != 0)
+            return {};
+
         if (url.scheme != "git" &&
             url.scheme != "git+http" &&
             url.scheme != "git+https" &&
             url.scheme != "git+ssh" &&
-            url.scheme != "git+file") return {};
+            url.scheme != "git+file")
+            warn("URL scheme '%s' is non-standard and requires 'git-remote-%s'.", url.scheme, url.scheme.substr(4, url.scheme.length() - 4));
```

Comment:
This should come with a way to turn off the warning, for users who consistently use this.
I do think dependencies on remote helpers should be discouraged, because they do make expressions that use rely on them harder to use, so enabling the warning by default would be good.

Reply:
This applies to basically flake inputs and `fetchTree`. We could also silence it automatically when a helper is requested by any installable on the CLI.

Reply:
> This applies to basically flake inputs and `fetchTree`. We could also silence it automatically when a helper is requested by any installable on the CLI.

This sounds like a terrible hack for something that may as well be two separate code paths, see https://github.com/NixOS/nix/pull/10567#issuecomment-2122560256

Reply:
I was still trying to figure out requirements, not describing an implementation.

The hack is of your inference. I don't think we need anything truly new architecturally.
The warning could be controlled by a global setting, that besides being a normal setting is set to true by the CLI when it encounters an explicit need for remote helpers in its installables.

Do you really mean a literal extra code path? That sounds a lot heavier than a boolean that controls a warning.